### PR TITLE
Tweak UTF-8 string literal analyzer and code fix after language change

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes https://github.com/dotnet/roslyn/issues/61517

In response to the language spec change to make the type of `"ABC"u8` be a `ReadOnlySpan<byte>` rather than a `byte[]`, this updates the analyzer and code fix I wrote previously to:
* Not be part of code cleanup or VS options
* Append `.ToArray()` on the end of the string literals it creates
* Still be an analyzer so that we can still educate people about new language features

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -49,10 +49,12 @@ protected override void InitializeWorker(AnalysisContext context)
                 if (!context.Compilation.LanguageVersion().IsCSharp11OrAbove())
                     return;
 
+                if (context.Compilation.GetTypeByMetadataName(typeof(ReadOnlySpan<>).FullName!) is null)
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs:52]: I don't recall the details. But there was bugs previously due to the use of `GetTypeByMetadataName`.


```suggestion
                if (context.Compilation.GetBestTypeByMetadataName(typeof(ReadOnlySpan<>).FullName!) is null)
```

File: src/VisualStudio/CSharp/Impl/EditorConfigSettings/DataProvider/CodeStyle/CSharpCodeStyleSettingsProvider.cs [hunk 2]
```diff
@@ -152,7 +152,6 @@ private IEnumerable<CodeStyleSetting> GetExpressionCodeStyleOptions(AnalyzerConf
             yield return CodeStyleSetting.Create(CSharpCodeStyleOptions.PreferRangeOperator, description: ServicesVSResources.Prefer_range_operator, editorConfigOptions, visualStudioOptions, updaterService, FileName);
             yield return CodeStyleSetting.Create(CSharpCodeStyleOptions.ImplicitObjectCreationWhenTypeIsApparent, description: CSharpVSResources.Prefer_implicit_object_creation_when_type_is_apparent, editorConfigOptions, visualStudioOptions, updaterService, FileName);
             yield return CodeStyleSetting.Create(CSharpCodeStyleOptions.PreferTupleSwap, description: ServicesVSResources.Prefer_tuple_swap, editorConfigOptions, visualStudioOptions, updaterService, FileName);
-            yield return CodeStyleSetting.Create(CSharpCodeStyleOptions.PreferUtf8StringLiterals, description: ServicesVSResources.Prefer_UTF8_string_literals, editorConfigOptions, visualStudioOptions, updaterService, FileName);
```
Comment (Youssef1313) [inline src/VisualStudio/CSharp/Impl/EditorConfigSettings/DataProvider/CodeStyle/CSharpCodeStyleSettingsProvider.cs:155]: If analyzer is still keeping the option, I think we should keep it in editorconfig UI.

Comment (davidwengier) [reply src/VisualStudio/CSharp/Impl/EditorConfigSettings/DataProvider/CodeStyle/CSharpCodeStyleSettingsProvider.cs:155]: Yeah, I got carried away, as indicated by the test failures. Adding this back now.

**Review by Youssef1313 — COMMENTED** [review]
An array is [implicitly convertible](https://docs.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.op_implicit?view=net-6.0#system-readonlyspan-1-op-implicit(t())-system-readonlyspan((-0))) to a `ReadOnlySpan<T>`
The `.ToArray` call *may* be unnecessary if the variable, for example, is passed to a method that accepts `ReadOnlySpan<T>` and the implicit conversion is used, or if the method has both overloads.

This is out of scope for this PR, but it might be beneficial to consider. I'm not sure what would be the best design for this. A separate simplifier that's always applied after code fixes are invoked? An annotation (similar to `AddImportsAnnotation`)? A separate analyzer/codefix?

File: src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs [hunk 3]
```diff
@@ -175,21 +175,26 @@ private static SyntaxNode CreateArgumentListWithUTF8String(BaseArgumentListSynta
             return argumentList.WithArguments(SyntaxFactory.SeparatedList<ArgumentSyntax>(arguments));
         }
 
-        private static LiteralExpressionSyntax CreateUTF8String(SyntaxNode nodeToTakeTriviaFrom, string stringValue)
+        private static InvocationExpressionSyntax CreateUTF8String(SyntaxNode nodeToTakeTriviaFrom, string stringValue)
         {
             return CreateUTF8String(nodeToTakeTriviaFrom.GetLeadingTrivia(), stringValue, nodeToTakeTriviaFrom.GetTrailingTrivia());
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs:180]:
```suggestion
            return CreateUTF8String(stringValue).WithTriviaFrom(nodeToTakeTrivia);
```

File: src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs [hunk 4]
```diff
@@ -175,21 +175,26 @@ private static SyntaxNode CreateArgumentListWithUTF8String(BaseArgumentListSynta
...
         }
 
-        private static LiteralExpressionSyntax CreateUTF8String(SyntaxTriviaList leadingTrivia, string stringValue, SyntaxTriviaList trailingTrivia)
+        private static InvocationExpressionSyntax CreateUTF8String(SyntaxTriviaList leadingTrivia, string stringValue, SyntaxTriviaList trailingTrivia)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs:183]: you can remove all trivia logic from this.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs (hunk 4 above)
Comment (davidwengier) [reply src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs:183]: Unfortunately, not. This is called for the param array case which needs to pass in no leading and trailing trivia from the last argument.

File: src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs (hunk 3 above)
Comment (davidwengier) [reply src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs:180]: As per below, switching this to use `WithTriviaFrom` would work, but the method would have to pass two empty trivia lists for the leading and trailing trivia arguments, so not worth it.

Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/CodeFixes/UseUTF8StringLiteral/UseUTF8StringLiteralCodeFixProvider.cs:180]: Thanks!!

**Pull request merged**